Userland/Libraries/LibCore/Stream.h

```diff
@@ -705,85 +684,81 @@ class BufferedHelper {
             }
         }
 
-        auto readable_size = min(m_buffered_size, buffer.size());
+        auto const candidate = TRY(populate_until_any_of(candidates));
+
+        if (candidate.has_value()) {
+            auto const buffer_written_for_end_user = m_buffer.read(buffer.slice(0, candidate->offset));
+            TRY(m_buffer.discard(candidate->match_size));
+            return buffer_written_for_end_user;
+        }
 
-        // The intention here is to try to match all of the possible
+        // If we still haven't found anything, then it's most likely the case
+        // that the delimiter ends beyond the length of the caller-passed
+        // buffer. Let's just fill the caller's buffer up.
+        return m_buffer.read(buffer);
+    }
+
+    struct OffsetAndSize {
+        size_t offset {};
+        size_t match_size {};
+    };
+
+    template<size_t N>
+    ErrorOr<Optional<OffsetAndSize>> populate_until_any_of(Array<StringView, N> const& candidates)
```

Comment:
Of course!
I will try to extract a refactoring commit and get it merged first, so it can minimize the diff here.

Reply:
This looks pretty manageable: fcb723304109aad31fdc1896f515104c88e1e2dc.

Reply:
Honestly, it's better than I thought, but it has *regressions* :scream:

#16544 factorize both usage of `offset_off` and actually introduce the small overhead needed to be compatible with the `CircularBuffer` API.

But it's not that big, so let's merge this first to avoid the rebase.

Reply:
I implemented `offset_off(StringView needle, Optional<size_t> until = {})` and added a test for it.
At least, the call to `offset_off` replacing `memmem` won't go through the whole buffer.